services/facility_queue/facility_queue.py

```diff
@@ -0,0 +1,195 @@
+import json
+import requests
+
+from sqlalchemy.orm import sessionmaker, scoped_session
+import sqlalchemy as sa
+
+from baselayer.app.models import init_db
+from baselayer.app.env import load_env
+from skyportal.models import (
+    DBSession,
+    FollowupRequest,
+    FacilityTransactionRequest,
+)
+
+env, cfg = load_env()
+
+init_db(**cfg['database'])
+
+request_session = requests.Session()
+request_session.trust_env = (
+    False  # Otherwise pre-existing netrc config will override auth headers
+)
+
+# class FacilityTransactionRequest:
+#    def __init__(self, method=None, endpoint=None, headers=None):
+#        self.method = method
+#        self.endpoint = endpoint
+#        self.headers = headers
+
+Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
+
+
+class FacilityQueue(asyncio.Queue):
+    async def load_from_db(self):
+        # Load items from database into queue
+
+        session = Session()
+        requests = (
+            session.query(FacilityTransactionRequest)
+            .where(FacilityTransactionRequest.status != "complete")
+            .all()
+        )
+        for req in requests:
+            await self.put(req)
+
+    async def service(self):
+        while True:
+            req = await queue.get()
+            print(f"{req.method} request to [{req.endpoint}]")
+
+            session = Session()
+            followup_request = session.query(FollowupRequest).get(
+                req.followup_request_id
+            )
+            instrument = followup_request.allocation.instrument
+            altdata = followup_request.allocation.altdata
+
+            response = request_session.request(
+                req.method,
+                req.endpoint,
+                json=req.data,
+                params=req.params,
+                headers=req.headers,
+            )
+
+            if instrument.name == "ATLAS":
+                from skyportal.facility_apis.atlas import commit_photometry
+
+                if response.status_code == 200:
+                    try:
+                        json_response = response.json()
+                    except Exception:
+                        raise ('No JSON data returned in request')
+
+                    if json_response['finishtimestamp']:
+                        req.status = "Committing photometry to database"
+                        commit_photometry(
+                            json_response,
+                            altdata,
+                            followup_request.id,
+                            instrument.id,
+                            followup_request.requester.id,
+                        )
+                        req.status = 'complete'
+                        session.add(req)
+
+                    elif json_response['starttimestamp']:
+                        followup_request.status = f"Task is running (started at {json_response['starttimestamp']})"
+                        await self.put(req)
```

Comment:
This means that, if the queue is empty and there is only one request, it will keep pinging the remote server.  Maybe check the difference between current time and last time tried, and ensure that is longer than, what, 2 minutes or so?

Reply:
Added!